### PR TITLE
URL-Sharing of Datasets

### LINKED ISSUE #2435: URL sharing of datasets
Create a shareable link for dataset viewing (like private youtube links)

<a href="https://timer.scm.io/repos/537237ce1a00001a003c4a14/issues/2435/create?referer=github" target="_blank">Log Time</a>

### PR DESCRIPTION
**Pull request by fm3**
### Mailable description of changes:
 - Private Datasets can be shared with anonymous users with a designated token

### Changes Backend
 * new optional parameter for `/api/datasets/:dataSetName`: `sharingToken`
 * new routes `GET` + `DELETE`  `/api/datasets/:dataSetName/sharingToken` (get will create new if none is present)

this sharingToken can also be sent to the datastore (just use the already existing `token`-field there).
Note: the sharingToken will only be used if the user is not logged in.

### Steps to test:
- Go to edit dataset view
- Generate and copy the sharing link
- logout (maybe turn off auto-login)
- try to open the sharing link (should work)
- try to open with with a different token (shouldn't work) 

### Issues:
- fixes #2435

------
- [x] Ready for review

### CODE REVIEW

**Comment by normanrz** [comment]
@fm3 Could we have shorter tokens? Something like 10 chars [a-z0-9]

**Comment by fm3** [comment]
@normanrz why, yes :)

**Comment by normanrz** [comment]
Frontend code is ready for review

**Comment by fm3** [comment]
so is the backend code. 
I’ll add the sql evolution tomorrow

File: app/assets/javascripts/dashboard/dataset/dataset_import_view.js [hunk 1]
```diff
@@ -114,13 +129,39 @@ class DatasetImportView extends React.PureComponent<Props, State> {
...
+
   updateDataset(propertyName: string, value: string | boolean) {
     const newState = update(this.state, {
       dataset: { [propertyName]: { $set: value } },
     });
     this.setState(newState);
   }
 
+  getSharingLink() {
+    return `${window.location.origin}/datasets/${this.props.datasetName}/view?token=${
```
[the remark this answers is not in the document]
Comment (normanrz) [reply app/assets/javascripts/dashboard/dataset/dataset_import_view.js]: This one creates the user-facing link. The other one is for accessing the API. Is that what you meant?

File: app/assets/javascripts/admin/admin_rest_api.js [hunk 2]
```diff
@@ -61,12 +61,21 @@ function requestUserToken(): Promise<string> {
   return tokenRequestPromise;
 }
 
+export function getSharingToken(): ?string {
+  if (window.location != null) {
+    const params = Utils.getUrlParamsObject();
+    if (params != null && Object.prototype.hasOwnProperty.call(params, "token")) {
```
[the remark this answers is not in the document]
Comment (normanrz) [reply app/assets/javascripts/admin/admin_rest_api.js]: You're right `params.token != null` is simpler

**Comment by daniel-wer** [comment]
I did some manual testing, the feature works very well :ship:

File: app/assets/javascripts/dashboard/dataset/dataset_import_view.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (normanrz) [reply app/assets/javascripts/dashboard/dataset/dataset_import_view.js]: Ah! Good catch that flow prop needs to go.

**Pull request merged**